### PR TITLE
fix: broken memcached error reporting

### PR DESCRIPTION
**Pull request by kostasrim**
Partially solves #1730 (because I suspect there is some other issue as well -- either the user it doing something illegal with the replica or there is some other error which is not reported properly because of this bug).

`ConnectionContext` overloads `operator->` by returning a `RedisReplyBuilder`. This is extremely error prone, because there are execution paths that work both with `memcached` and `redis` protocol. If used improperly in a memcached execution path it will trigger an internal assertion `Protocol::REDIS == protocol()` instead of properly report the error to the client.

This PR addresses this, by adding the `SendError` family of functions in `ConnectionContext`. I replace the `(*cntx)->` with `cntx->SendError` which uses the internal `SinkReplyBuilder` and dispatching to the right function based on the underline protocol. 

All in all this PR fixes:

* DispatchCommand error reporting (one example is when we use SET command on the replica -- previously we crashed now we properly report an error)
* INCRBY command error reporting (this would crash)
* DECRBY command error reporting (this would crash)

Basically, I went over all of the memcached commands just to make sure that we report errors properly via the common base client `SinkReplyBuilder` and not via the specific `RedisReplyBuilder` class via the overloaded arrow operator.

I really don't like the `operator->` as it's extremely error prone and it's really hard to assert that an execution path is redis only (especially when both `->` and `reply_builder` are used within the same function -- it makes you scratch your head if it's a bug or an actual feature).

Generally speaking a long term solution is to replace all of the common functions to strictly use a common interface via `SinkReplyBuilder` such that are protocol agnostic (since anyway they should be used by both memcached and redis)

### CODE REVIEW

File: src/facade/reply_builder.cc [hunk 1]
```diff
@@ -277,7 +289,7 @@ void RedisReplyBuilder::SendBulkString(std::string_view str) {
   return Send(v, ABSL_ARRAYSIZE(v));
 }
 
-std::string_view RedisReplyBuilder::StatusToMsg(OpStatus status) {
+std::string_view SinkReplyBuilder::StatusToMsg(OpStatus status) {
```
Comment (romange) [inline src/facade/reply_builder.cc:292]: why did you change it? that's not good because these errors are not part of memcached protocol.

**Comment by romange** [comment]
I think we should add a pytest that reproduces the problem.

File: src/facade/reply_builder.cc (hunk 1 above)
Comment (kostasrim) [reply src/facade/reply_builder.cc:292]: That's not true: 

```
   949     if (cid->IsTransactional()) {       
   950       dist_trans.reset(new Transaction{cid});                                                     
   951                
   952       if (!dist_trans->IsMulti()) {  // Multi command initialize themself based on their mode.
   953         if (auto st = dist_trans->InitByArgs(dfly_cntx->conn_state.db_index, args_no_cmd);        
   954             st != OpStatus::OK)                
   955           return cntx->SendError(st);
   956       }                                                                                           
   957                                                                    
   958       dfly_cntx->transaction = dist_trans.get();
   959       dfly_cntx->last_command_debug.shards_count = dfly_cntx->transaction->GetUniqueShardCnt();   
   960     } else {                                                                                      
   961       dfly_cntx->transaction = nullptr;                                                           
   962     }                                                                        
   963   }  
```

Andddd 

`<< CI{"SET", CO::WRITE | CO::DENYOOM | CO::NO_AUTOJOURNAL`

So memcached commands are transactional. Same applies for `get` `incr` `decr` etc. If the weren't, our transactional system would be broken, because memcached commands would not be a part of a transaction whereas redis commands would

File: src/facade/conn_context.h [hunk 2]
```diff
@@ -50,6 +52,22 @@ class ConnectionContext {
...
+    rbuilder_->SendError(std::move(error));
+  }
+
+  void SendError(OpStatus status) {
+    rbuilder_->SendError(status);
+  }
+
+  void SendSimpleString(std::string_view str) {
+    rbuilder_->SendSimpleString(str);
+  }
```
Comment (dranikpg) [inline src/facade/conn_context.h:69]: One day I'll make ErrorReply's constructor implicit and remove all the overloads 😆

File: src/facade/reply_builder.cc (hunk 1 above)
Comment (romange) [reply src/facade/reply_builder.cc:292]: `InitByArgs` always succeeds for these commands.

Comment (romange) [reply src/facade/reply_builder.cc:292]: We basically validate everything for memcached already at the parser level

**Comment by romange** [comment]
@kostasrim  I added this test to the main branch (locally on my machine) and nothing has crashed.

**Comment by kostasrim** [comment]
@romange Yes because I forgot to check in the changes in `__init__`. We have another bug there:

```
105     @property
106     def mc_port(self) -> int:
107         if self.params.existi get(key: _KT) -> Optional[_VT_co]                              
108             return self.param get(key: _KT, default: Union[_VT_co, _T]) -> Union[_VT_co, _T] 
109         return int(self.args.get("mc_port", "11211"))
```
It works because you are not connecting to the replica. Will push in 5 and you should now get a nice error in main:

`pymemcache.exceptions.MemcacheUnexpectedCloseError` which is basically the replica crashing....

File: src/facade/reply_builder.cc (hunk 1 above)
Comment (kostasrim) [reply src/facade/reply_builder.cc:292]: really bug prone tbh, some of the error reporting is already handled while some other doesn't(in case of this bug). Anyway, beyond the scope of this PR. I will see at some point how we can simplify this.

Comment (kostasrim) [reply src/facade/reply_builder.cc:292]: I removed this for now

Comment (kostasrim) [reply src/facade/reply_builder.cc:292]: actually no, we kinda need this because we need `SendError(ErrorReply)` which uses `status`. 

to be fair, it won't ever arrive to that block that uses status so I ended up removing this and send an internal error if the `ErrorReply` contains a status. You never know if we hit that so at least let's report this to be sure.

File: tests/dragonfly/__init__.py [hunk 3]
```diff
@@ -106,7 +106,7 @@ def admin_port(self) -> int:
     def mc_port(self) -> int:
         if self.params.existing_mc_port:
             return self.params.existing_mc_port
-        return int(self.args.get("mc_port", "11211"))
+        return int(self.args.get("memcached_port", "11211"))
```
Comment (kostasrim) [inline tests/dragonfly/__init__.py:109]: @romange this is why it was passing....

File: src/server/string_family.cc [hunk 4]
```diff
@@ -976,7 +976,7 @@ void StringFamily::IncrBy(CmdArgList args, ConnectionContext* cntx) {
   int64_t val;
 
   if (!absl::SimpleAtoi(sval, &val)) {
-    return (*cntx)->SendError(kInvalidIntErr);
+    return cntx->SendError(kInvalidIntErr);
```
Comment (romange) [inline src/server/string_family.cc:979]: i know it looks fragile but memcache parser already does all the parsing checks because those conditions should not trigger with memcache.

File: src/facade/reply_builder.h [hunk 5]
```diff
@@ -57,6 +58,10 @@ class SinkReplyBuilder {
 
   virtual void SendProtocolError(std::string_view str) = 0;
 
+  // You normally should not call this - maps the status
+  // into the string that would be sent
+  static std::string_view StatusToMsg(OpStatus status);
```
Comment (romange) [inline src/facade/reply_builder.h:63]: It's not your fault but those local fixes make the whole thing more confusing.
Let's do the following:
1. Move StatusToMsg to op_status.h as ` std::string_view StastusMessage(OpStatus status);`
2. let's make `SendError(ErrorReply error)` non virtual method in SinkReplyBuilder. 

I also changed how SendError is implemented for MC reply builder, I will send you a PR.

File: src/facade/conn_context.h (hunk 2 above)
Comment (romange) [reply src/facade/conn_context.h]: it could be nice.

File: src/facade/reply_builder.h (hunk 5 above)
Comment (kostasrim) [reply src/facade/reply_builder.h:63]: > let's make SendError(ErrorReply error) non virtual method in SinkReplyBuilder.

We can't really do that because of  `void CapturingReplyBuilder::SendError(ErrorReply error)` actually overrides it with a different implementation.

Comment (dranikpg) [reply src/facade/reply_builder.h]: We overwrites the others as well, it just becomes a little less efficient because it can't move out of the string IIRC

File: src/facade/reply_builder.cc [hunk 6]
```diff
@@ -94,6 +94,13 @@ void SinkReplyBuilder::SendRaw(std::string_view raw) {
   Send(&v, 1);
 }
 
+void SinkReplyBuilder::SendError(ErrorReply error) {
+  DCHECK(!error.status);
+
+  string_view message_sv = visit([](auto&& str) -> string_view { return str; }, error.message);
```
[the remark this answers is not in the document]
Comment (romange) [reply src/facade/reply_builder.cc:100]: and since you missed, please add this case to reply_builder_test :)

File: src/facade/reply_builder_test.cc [hunk 7]
```diff
@@ -251,14 +251,39 @@ TEST_F(RedisReplyBuilderTest, ErrorBuiltInMessage) {
   }
 }
 
+TEST_F(RedisReplyBuilderTest, ErrorReplyBuiltInMessage) {
```
Comment (kostasrim) [inline src/facade/reply_builder_test.cc:254]: @romange This is a little "raw" but I wanted to verify that both of the execution paths are covered with `ErrorReply` without the extra nuances of printing extra error information through the `<<` operator. Moreover, it doesn't cover all sorts possible `OpStatus` values and the `e1`, `e2` are just placeholders such that I can verify that the right execution path got triggered

**Review by romange — APPROVED** [review]
